File: TypeScript/src/devicedriver.ts
import { nanoseconds as nanoTime } from "nano-time";
import { FlashMemoryDevice } from "./flashmemorydevice";

const INIT_ADDRESS = 0x00;

const PROGRAM_COMMAND = 0x40;
const READY_MASK = 0x02;
const READY_NO_ERROR = 0x00;

const TIMEOUT_THRESHOLD = 100000000;

const RESET_COMMAND = 0xFF;
const VPP_MASK = 0x20;
const INTERNAL_ERROR_MASK = 0x10;
const PROTECTED_BLOCK_ERROR_MASK = 0x08;

/**
 * This class is used by the operating system to interact with the hardware.
 */
export class DeviceDriver {
    private hardware: FlashMemoryDevice;

    constructor(hardware: FlashMemoryDevice) {
        this.hardware = hardware;
    }

    public read(address: number): number {
        return this.hardware.read(address);
    }

    public write(address: number, data: number): void {
        const start = nanoTime();
        this.hardware.write(INIT_ADDRESS, PROGRAM_COMMAND);
        this.hardware.write(address, data);
        let readyByte;
        while (((readyByte = this.read(INIT_ADDRESS)) & READY_MASK) === 0) {
            if (readyByte !== READY_NO_ERROR) {
                this.hardware.write(INIT_ADDRESS, RESET_COMMAND);
                if ((readyByte & VPP_MASK) > 0) {
                    throw new VppException();
                }
                if ((readyByte & INTERNAL_ERROR_MASK) > 0) {
                    throw new InternalErrorException();
                }
                if ((readyByte & PROTECTED_BLOCK_ERROR_MASK) > 0) {
                    throw new ProtectedBlockException();
                }
            }
            if (nanoTime() - start > TIMEOUT_THRESHOLD) {
                throw new TimeoutException("Timeout when trying to read data from memory");
            }
        }
        const actual = this.read(address);
        if (data !== actual) {
            throw new ReadFailureException("Failed to read data from memory");
        }
    }
}

export class VppException extends Error {
    constructor() {
        super("vpp exception");
    }
}

export class InternalErrorException extends Error {
    constructor() {
        super("internal error");
    }
}

export class ProtectedBlockException extends Error {
    constructor() {
        super("protected block");
    }
}

export class TimeoutException extends Error {
    constructor(message: string) {
        super(message);
    }
}

export class ReadFailureException extends Error {
    constructor(message: string) {
        super(message);
    }
}

File: Python/device_driver.py
import time


class VppException(Exception):
    pass


class InternalErrorException(Exception):
    pass


class ProtectedBlockException(Exception):
    pass


class ReadFailureException(Exception):
    def __init__(self, message):
        super().__init__(message)


class TimeoutException(Exception):
    def __init__(self, message):
        super().__init__(message)

class FlashMemoryDevice:
    """
    This class represents the interface to a Flash Memory Device. The hardware has only two methods - 'read' and 'write'
    However, the interface for using the device is a lot more complex than that. It is outlined in the top-level README file.
    """
    def read(self, address: int) -> int:
        raise NotImplementedError("unsupported operation")

    def write(self, address: int, data: int):
        raise NotImplementedError("unsupported operation")

class DeviceDriver:
    """
    This class is used by the operating system to interact with the hardware 'FlashMemoryDevice'.
    """

    INIT_ADDRESS = 0x00

    PROGRAM_COMMAND = 0x40
    READY_MASK = 0x02
    READY_NO_ERROR = 0x00

    TIMEOUT_THRESHOLD = 100_000_000  # nanoseconds

    RESET_COMMAND = 0xFF
    VPP_MASK = 0x20
    INTERNAL_ERROR_MASK = 0x10
    PROTECTED_BLOCK_ERROR_MASK = 0x08

    def __init__(self, hardware: FlashMemoryDevice):
        self.hardware = hardware

    def read(self, address: int) -> int:
        return self.hardware.read(address)

    def write(self, address: int, data: int):
        start = time.perf_counter_ns()
        self.hardware.write(self.INIT_ADDRESS, self.PROGRAM_COMMAND)
        self.hardware.write(address, data)

        while True:
            ready_byte = self.read(self.INIT_ADDRESS)
            if ready_byte & self.READY_MASK:
                break

            if ready_byte != self.READY_NO_ERROR:
                self.hardware.write(self.INIT_ADDRESS, self.RESET_COMMAND)

                if ready_byte & self.VPP_MASK:
                    raise VppException()

                if ready_byte & self.INTERNAL_ERROR_MASK:
                    raise InternalErrorException()

                if ready_byte & self.PROTECTED_BLOCK_ERROR_MASK:
                    raise ProtectedBlockException()

            if time.perf_counter_ns() - start > self.TIMEOUT_THRESHOLD:
                raise TimeoutException("Timeout when trying to read data from memory")

        actual = self.read(address)
        if data != actual:
            raise ReadFailureException("Failed to read data from memory")
